[#43684] migrated op-modal to spot-modal

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/43684
- amended styles
- removed "modal--title"
- removed header component
- removed closing icon
- fixed backdrop click
- removed jquery from modal service

TODO:
- [x] mobile behavior
- [x] fix feature spec selectors

Kharonus committed:
[#43684] migrated op-modal to spot-modal

- https://community.openproject.org/work_packages/43684
- amended styles
- removed "modal--title"
- removed header component
- removed closing icon
- fixed backdrop click
- removed jquery from modal service

Kharonus committed:
Merge branch 'dev' into implementation/43684-add-modal-dialog-to-design-system

Kharonus committed:
[#43684] removed unused styling

- fixed some inconsitencies
- readded removed icon
- removed listener on close button

Kharonus committed:
[#43684] combined spot-action-bar

- used spot-action-bar in combination with spot-modal--footer
- fixed some styling in action bar

Kharonus committed:
[#43684] replaced more op-modal

- moved spec selectors to spot-modal
- replaced found op-modal in html templates

Kharonus committed:
[#43684] add no margin to old buttons in new dialogs

- added mobile modal behavior
- fixed some feature spec selectors

HDinger commented on app/views/layouts/base.html.erb:
```diff
@@ -117,7 +117,7 @@ See COPYRIGHT and LICENSE files for more details.
         <section data-augmented-model-wrapper
                  data-modal-initialize-now="true"
-                 data-modal-class-name="onboarding-modal op-modal_autoheight -highlight">
+                 data-modal-class-name="onboarding-modal spot-modal_autoheight -highlight">
```
The `-highlight` class can be removed. It is also still present in some other files.

Kharonus replied:
removed usage of `-highlight` and `spot-modal_autoheight`. Second is fine, as the standard modal now has a calculated height.

HDinger commented on app/views/layouts/base.html.erb:
```diff
@@ -117,7 +117,7 @@ See COPYRIGHT and LICENSE files for more details.
         <section data-augmented-model-wrapper
                  data-modal-initialize-now="true"
-                 data-modal-class-name="onboarding-modal op-modal_autoheight -highlight">
+                 data-modal-class-name="onboarding-modal spot-modal_autoheight -highlight">
```
The class ` spot-modal_autoheight` does not exist. I assume it should be either removed completely, or be just `spot-modal`?

Kharonus replied:
as far as I understand it, this one receives the `spot-modal` through `data-modal-initialize-now`. otherwise it would now have worked. I try to find the usage of that one.

Kharonus replied:
well, the `spot-modal` is within the partial rendered in there.

HDinger commented on frontend/src/app/core/setup/globals/onboarding/onboarding_tour_trigger.ts:
```diff
@@ -33,14 +33,6 @@ export function detectOnboardingTour():void {
       currentTourPart = '';
       sessionStorage.setItem(onboardingTourStorageKey, 'readyToStart');
 
-      waitForElement('.onboarding-modal .op-modal--close-button', 'body', () => {
```
This should probably be replaced by a listener to an outside click.

Kharonus replied:
True, replaced it. good catch! 👍🏽

HDinger commented on frontend/src/app/spot/styles/sass/components/modal-overrides.sass:
```diff
@@ -1,7 +1,10 @@
 // augmented modals are loaded, this is not so easy to change.
 // TODO: When refactoring the modal service, this should be removed
-.op-modal > .form:only-child
+.spot-modal > .form:only-child
```
I am not so happy about placing that inside spot, as we don't want legacy code here. Maybe @b12f has a strong opinion on where to put that.

b12f replied:
Maybe putting it in these `$block-overrides.sass` files is the way to go @HDinger. We could do the same for the `spot-input` where foundation is in the way.

Kharonus replied:
what is the resulting suggestion? I do not get it yet.

Kharonus replied:
as discussed, leaving it here.

HDinger commented on frontend/src/global_styles/content/_modal.lsg:
````diff
@@ -81,29 +77,32 @@
 </div>
 ```
 
 ## Modals: Danger zone highlighting
````
Thanks for also updating the LSG 👍  This whole block can be removed, as there is no such thing as a danger modal any more.

Kharonus replied:
removed the highlighted one, but kept the `danger zone` to indicated the button style matter

also added a close footer to the standard one.

HDinger commented on modules/auth_plugins/app/views/hooks/login/_providers_css.html.erb:
```diff
@@ -36,7 +36,7 @@ See COPYRIGHT and LICENSE files for more details.
         background-image:  url('<%= asset_path(pro[:icon]) %>') ;
       }
-      .op-modal .login-auth-providers a.auth-provider.auth-provider-<%= pro[:name] %> {
+      .spot-modal .login-auth-providers a.auth-provider.auth-provider-<%= pro[:name] %> {
```
I'd like to avoid the `spot-modal` class to be overwritten somewhere outside. Please add a custom class to that modal, which you can reuse here.

Kharonus replied:
selector without `spot-modal` is strong enough at this place. Removed `.spot-modal`

HDinger reviewed (changes requested):
Great Job, @Kharonus ! I did not test all modals since there are simply too many. However, the samples were looking good 👍 

I have only some minor remarks:

- [x] There are some references to the op-modal in sass-files
- [x] The scrollbar is not the styled scrollbar 
- [ ] The subscriptions plugin is missing. ( I opened a separate ticket for that: )

There are two things I am missing, which needs to be decided by @psatyal or @marcalcobe however:

* For me it is not intuitive to click outside to close the modal. I was actively looking for either the cross or a "close" button
* Just showing a red button in a deletion modal is not striking enough in my eyes. I have a red color scheme enabled. For me, all buttons are red. So these modal don't really warn me about a dangerous action.

marcalcobe commented:
Hi @HDinger,

I haven't talked with @psatyal yet but my thoughts on the two things you mentioned:

- The close by clicking outside of the modal (black 50% area) is a really extended and known but in any case its only the secondary option to close the modal. The main option is always present in the action bar of the modal as a "Cancel" or "Close" button.
- About the red button I think this is a support to show a destructive action, so for me the relevance is that it comes in a full screen modal (this should already alert the user) and then there is a red button on it. On the other hand, unfortunately the red colour is socially associated to warning, destructive or restrictive actions so it is true that in red based themes it loses prominence but socially it is the correct to use in this case.

Kharonus committed:
[#43684] added modal close on mobile resolutions

HDinger commented:
Hi @marcalcobe 

> The close by clicking outside of the modal (black 50% area) is a really extended and known but in any case its only the secondary option to close the modal. The main option is always present in the action bar of the modal as a "Cancel" or "Close" button.

Unfortunatley, there is not always a "close" or "cancel" button in the modals at the moment. So we'd either have to add them now, or think of an alternative solution.

Kharonus commented:
Backdrop clicks are quite common, I agree.

Just need to raise the alert, that there will be modals (old ones) that do not have a cancel button in the action bar (some of those do not have any button, like the export modals.

marcalcobe commented:
After a discussion with @psatyal we decided that the headers will remain without colour. We understand that might be less clear for users who are using a red theme but from a design perspective we want to keep the modals clean for the majority of users.

About the close action on the action bar the idea is that all the modals will now contain action bar, I will create a work package so this is clear as we discussed.

Kharonus commented:
the ticket for the refactoring of the op-modals in the subscription plugin is: https://community.openproject.org/projects/openproject/work_packages/43822/activity

Kharonus committed:
[#43684] fixed PR issues

- removed old css classes from html
- fixed feature spec click
- readded tour trigger on modal cancel

bsatarnejad committed:
when clicking outside the modal it will be closed instaed of clicking on a close button

bsatarnejad committed:
Merge branch 'dev' into implementation/43684-add-modal-dialog-to-design-system

b12f commented on frontend/src/app/spot/styles/sass/components/modal.sass:
```diff
@@ -0,0 +1,80 @@
+      outline-style: none
+
+    &_limited-height
+      max-height: 50vh
```
If this isn't part of the specification, I'd remove it

b12f replied:
We need to keep sure it's limited so all modals stay within 100vh though

bsatarnejad replied:
All modals have the max-width and max-height set in spot-modal class, though they won't get higher or wider than the screen. So for now, I think we can remove this class (that is used for attribute help text modal) and wait for the final specifications of each modal.

b12f commented on frontend/src/app/spot/styles/sass/components/modal.sass:
```diff
@@ -0,0 +1,80 @@
+
+  &_fullscreen
+    width: 100vw
+    height: 100vh
```
I don't think this is in use at all, it isn't part of the specification either way I think

b12f reviewed (changes requested):
In the name of correctness, I think we should keep our CSS as close to the design spec as we can

b12f commented on frontend/src/app/features/invite-user-modal/success/success.component.html:
```diff
@@ -1,5 +1,5 @@
 <div class="op-ium-success">
-  <div class="op-modal--body op-modal--body_center">
+  <div class="spot-modal--body spot-modal--body_center">
```
The `_center` class should be a special override for just this component for now

bsatarnejad committed:
exclude _center class from spot-modal sass file itself and add it to the component style sheet

bsatarnejad committed:
Words should be break in any case like when there is no space there

bsatarnejad committed:
remove unnecessary classes that never used like _fullscreen and _cente, remove _limited-height class to make all the modal look the same

Commit:
Merge branch 'dev' into implementation/43684-add-modal-dialog-to-design-system

b12f reviewed (approved):
LGTM, really nice work!

oliverguenther commented on frontend/src/global_styles/content/user-content/_typography.sass:
```diff
@@ -43,6 +43,7 @@
   margin: 0
   line-height: 1.6em
   color: inherit
+  word-break: break-all
```
@Kharonus @bsatarnejad  this line causes https://community.openproject.org/projects/openproject/work_packages/44125/activity?query_id=3470 since it's applied to all paragraphs now.

Do you know where this originated from, and can we apply it to modal only?
